[Tablet Orders] Update selected products and product selector when changing size class

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11923
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

When the Product Selector is shown side-by-side with the Order Form on tablets and larger iPhones, selections are synced to the order immediately. When shown in a modal-on-modal context, the merchant has to tap a button to dismiss the selector before we sync their changes.

Both iPads and large iPhones can switch between these display approaches, by rotating the device or moving to a split view/slideover presentation.

Previously, when switching from modal-on-modal to side-by-side with the product selector visible, any unsynced selection changes would be lost. Additionally, the product selector would show an titleless button, and the merchant would have to tap that to sync their selections.

This PR fixes these issues, so any unsynced changes are saved to the order when we move from modal-on-modal to side-by-side.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Enable the `sideBySideViewForOrderForm` feature flag
2. Launch the app on an iPad or large iPhone (in landscape)
3. Open the `Orders` tab, and then tap `+` to make a new order – it should be in side-by-side mode
4. Select a product
5. Rotate the iPhone or set a split view to put the order form in modal-on-modal mode
6. Tap `+` to open the product selector
7. Change your selections, but don't tap the button
8. Rotate the iPhone or expand the split view to put the screen in side-by-side mode
9. Observe that your changes are synced, not lost.

Repeat as closely as you can with the feature flag disabled – e.g. by closing the product selector, and looking for products that have been unexpectedly synced – there should be no change in behaviour.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/woocommerce/woocommerce-ios/assets/2472348/4388d4c3-564f-4954-b7d8-1b1973e793a6


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
11923 Simplify product selector configuration

joshheald committed:
11923 Use size class to set selected sync approach

The selected products are synced immediately when the product selector is shown in a split view, or on tapping a button when it’s shown as a modal.

This commit expresses these two approaches as an enum, and sets them based on the size class changing.

I’ve opted not to use the `adaptiveModalContainerPresentationStyle` (i.e. `modalOnModal` or `sideBySide`, as it wasn’t triggering `onChange(of: perform:)` as it should have, and is a plain wrapper for `horizontalSizeClass` anyway. We are planning to remove it later this week.

joshheald committed:
11923 Only sync when moving to immediate

There’s no need to sync any changes when moving to `onButtonTap` – all the changes made while in `immediate` will have been synced, so adding another call to `onMultipleSelectionCompleted` only risks blocking the UI unneccesarily.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Fix whitespace

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -1404,6 +1404,70 @@ final class ProductSelectorViewModelTests: XCTestCase {
+        }
+
+        // Then
+        assertEqual([1], selectedItems)
```
By checking the assertion I initially thought this was an item count, but it's the selected item IDs. Wdyt if we rename it to:
```suggestion
        assertEqual([1], selectedItemIDs)
```

joshheald replied:
Done in 41f7880

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -1404,6 +1404,70 @@ final class ProductSelectorViewModelTests: XCTestCase {
+    func test_updateSyncApproach_doesnt_call_onMultipleSelectionCompleted_when_approach_stays_on_immediate() {
+        // Given
+        let expectation = XCTestExpectation(description: "onMultipleSelectionCompleted called")
+        expectation.isInverted = true
```
TIL `isInverted` , nice!

joshheald replied:
Yeah me too 😅 I just figured there must be something.

Best used sparingly, as it needs to wait for the full length of `timeout` to know that it hasn't happened. I've reduced that to 0.1s, which works in my testing, when I force it to fail. The tests take about 25% of that time to fail on my machine, including everything before the wait, so it should be long enough to allow CI to fail when appropriate as well. cd8a200

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:
```diff
@@ -121,34 +118,36 @@ struct OrderFormPresentationWrapper: View {
                         ConfigurableBundleProductView(viewModel: viewModel)
                     }
                 }
-            }, onSizeClassChange: { sizeClass in
-                productSelectorConfiguration = ProductSelectorConfiguration.loadConfiguration(for: sizeClass)
```
Thanks for updating this, I see now that, as you mentioned, the environment key is just a wrapper around the system-provided `horizontalSizeClass`, so we could have used that directly. If I had checked that initially, rather than focusing in our custom key, I could have avoided some pain 😅 

💯

iamgabrielma reviewed (approved):
Works well! 🚢

joshheald committed:
11923 Rename for clarity in sync approach test

joshheald committed:
11923 Reduce wait time for negative tests

These tests take about 0.02s when forced to fail on my machine, so 0.1s should give enough headroom on CI
